### PR TITLE
Fix Start-Process -Wait with -Credential

### LINKED ISSUE #17033: Start-Process -Wait doesn't work when running as a different user (-Credential)
### Prerequisites

- [X] Write a descriptive title.
- [X] Make sure you are able to repro it on the [latest released version](https://github.com/PowerShell/PowerShell/releases)
- [X] Search the existing issues.
- [X] Refer to the [FAQ](https://github.com/PowerShell/PowerShell/blob/master/docs/FAQ.md).
- [X] Refer to [Differences between Windows PowerShell 5.1 and PowerShell](https://docs.microsoft.com/powershell/scripting/whats-new/differences-from-windows-powershell).

### Steps to reproduce

On Windows, when you combine `Start-Process -Wait` with `-Credential` (running as a different user), the process is launched, BUT
* an `Access denied` error is reported
* no waiting occurs.

By contrast, using `-PassThru` and `.WaitForExit()` on the passed through `System.Diagnostics.Process` instance works fine.

```powershell
# Run on Windows
# Be sure to specify valid credentials when prompted.
Start-Process -Credential (Get-Credential) cmd  -Wait
```

### Expected behavior

No error should occur, and the call should block until the launched `cmd` instance is exited.

### Actual behavior

The `cmd` instance is launched, but, instead of waiting, an `Access denied` error is emitted.

Contrast this with this seemingly equivalent command, which works as expected:

```powershell
(Start-Process -Credential (Get-Credential) cmd -PassThru).WaitForExit()
```

### Error details

_No response_

### Environment data

```powershell
PowerShell Core 7.3.0-preview.2 on Windows
```

### Visuals

_No response_

### PR DESCRIPTION
**Pull request by jborean93**
# PR Summary

Allows a non-administrator user to be able to -Wait on a Start-Process call with a custom credential specified.

~~Currently thinking off a way I can test out this scenario.~~ After creating a test to run in CI I've found that it's too hard to create a reliable test that wouldn't block the whole runner on a failing scenario. The test itself also required admin access and was quite complicated to set up all the synchronisation work so I've decided not to include it.

Fixes: https://github.com/PowerShell/PowerShell/issues/17033
Alternative to https://github.com/PowerShell/PowerShell/pull/19082

## PR Context

https://github.com/PowerShell/PowerShell/issues/17033

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/main/reference/7.3/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): 
- **User-facing changes**
    - [x] Not Applicable
    - **OR**
    - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: 
- **Testing - New and feature**
    - [x] N/A or can only be tested interactively
    - **OR**
    - [ ] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [x] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and opened an issue in the relevant tool repository. This may include:
        - [ ] Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode
        (which runs in a different PS Host).
            - [ ] Issue filed: 
        - [ ] Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
            - [ ] Issue filed: 
        - [ ] Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
            - [ ] Issue filed: 
        - [ ] Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).
            - [ ] Issue filed:

### CODE REVIEW

File: src/Microsoft.PowerShell.Commands.Management/commands/management/Process.cs [hunk 1]
```diff
@@ -744,7 +744,8 @@ private static string RetrieveProcessUserName(Process process)
             try
             {
                 int error;
-                if (!Win32Native.OpenProcessToken(process.Handle, TOKEN_QUERY, out processTokenHandler)) { return null; }
+                if (!Win32Native.OpenProcessToken(process.Handle, TOKEN_QUERY, out processTokenHandler))
+                { return null; }
```
Comment (iSazonov) [inline src/Microsoft.PowerShell.Commands.Management/commands/management/Process.cs:748]: Too many unrelated changes in the PR. Please revert or move to another PR.

File: src/Microsoft.PowerShell.Commands.Management/commands/management/Process.cs [hunk 2]
```diff
@@ -2006,15 +2025,65 @@ protected override void BeginProcessing()
...
+
+                // Starting a process as another user might make it impossible
+                // to get the process handle from the S.D.Process object. Use
+                // the ALL_ACCESS token from CreateProcess here to setup the
+                // job object assignment early if -Wait was specified.
+                // https://github.com/PowerShell/PowerShell/issues/17033
+                if (Wait)
+                {
+                    jobObject = new();
+                    jobAssigned = jobObject.AssignProcessToJobObject(processInfo.Process.DangerousGetHandle());
```
Comment (iSazonov) [inline src/Microsoft.PowerShell.Commands.Management/commands/management/Process.cs:2080]: Can we use `process.SafeHandle.DangerousGetHandle`? If so we could do minimal changes.

File: src/Microsoft.PowerShell.Commands.Management/commands/management/Process.cs (hunk 1 above)
Comment (jborean93) [reply src/Microsoft.PowerShell.Commands.Management/commands/management/Process.cs:748]: Yea sorry that’s vscode autoformating files on save. I’ll revert them before making the PR ready.

File: src/Microsoft.PowerShell.Commands.Management/commands/management/Process.cs (hunk 2 above)
Comment (jborean93) [reply src/Microsoft.PowerShell.Commands.Management/commands/management/Process.cs:2080]: I can most likely have the AssingProcessToJobObject accept a SafeHandle and avoid the DangerousGetHand call. Ultimately it does lead to the same thing though.

File: src/Microsoft.PowerShell.Commands.Management/commands/management/Process.cs (hunk 1 above)
Comment (jborean93) [reply src/Microsoft.PowerShell.Commands.Management/commands/management/Process.cs]: These have been reverted with the latest commit.

File: src/Microsoft.PowerShell.Commands.Management/commands/management/Process.cs [hunk 3]
```diff
@@ -2014,7 +2014,55 @@ protected override void BeginProcessing()
...
+            if (startInfo.UseShellExecute)
+            {
+                process = StartWithShellExecute(startInfo);
+            }
+            else
+            {
+#if UNIX
+                process = new Process() { StartInfo = startInfo };
+                SetupInputOutputRedirection(process);
+                process.Start();
```
Comment (iSazonov) [inline src/Microsoft.PowerShell.Commands.Management/commands/management/Process.cs:2032]: Do you know anything stopping us from direct using process.Start() on Windows too?

I see only limitation - WindowStyle. .Net API support this only for shellexecution. We could ask .Net team to enhance the API and then we could remove the pinvoke at all.

Comment (jborean93) [reply src/Microsoft.PowerShell.Commands.Management/commands/management/Process.cs:2032]: It should be technically possible. I originally assumed the biggest blocker was that the dotnet API didn't support specifying the redirection file paths and the code would have to manually do all that. But it seems like it already does that for the non-Windows side so we could just reuse that. There could be other things that might crop up but I preferred to just address the original problem in question and potentially look at the larger change in the future.

Comment (iSazonov) [reply src/Microsoft.PowerShell.Commands.Management/commands/management/Process.cs:2032]: Thanks for confirmation! I will open new issue in .Net repo. If they confirm we can continue in more simple way.

Comment (iSazonov) [reply src/Microsoft.PowerShell.Commands.Management/commands/management/Process.cs:2032]: Isuue https://github.com/dotnet/runtime/issues/81681

File: src/Microsoft.PowerShell.Commands.Management/commands/management/Process.cs [hunk 4]
```diff
@@ -2014,7 +2014,55 @@ protected override void BeginProcessing()
...
+#else
+                using ProcessInformation processInfo = StartWithCreateProcess(startInfo);
+                process = Process.GetProcessById(processInfo.ProcessId);
+
+                // Starting a process as another user might make it impossible
+                // to get the process handle from the S.D.Process object. Use
+                // the ALL_ACCESS token from CreateProcess here to setup the
+                // job object assignment early if -Wait was specified.
+                // https://github.com/PowerShell/PowerShell/issues/17033
+                if (Wait)
```
Comment (SteveL-MSFT) [inline src/Microsoft.PowerShell.Commands.Management/commands/management/Process.cs:2056]: It's not clear to me why you need to assign the job object here rather than just do it once at line 2098?

**Review by SteveL-MSFT — APPROVED** [review]
Thanks for submitting this!  LGTM.  The changes here helped me understand what exactly you were asking for previously.

**Comment by jborean93** [comment]
@SteveL-MSFT is this waiting on anything?

**Review by iSazonov — APPROVED** [review]

**Comment by PaulHigin** [comment]
@jborean93   Thanks for your contribution!
